src/module_2/module_2_1.md

```diff
@@ -0,0 +1,402 @@
+#     missing_values = df_clean[column].isna()
+#     s = df_clean[column].value_counts(normalize=True)
+#     np.random.seed(99)
+#     df_clean.loc[missing_values, column] = np.random.choice(
+#         s.index, size=missing_values.sum(), p=s.values
+#     )
+
+# # Recalculating count_people column where np.nan
+# missing_values = df_clean["count_people"].isna()
+# df_clean.loc[missing_values, "count_people"] = (
+#     df_clean.loc[missing_values, "count_adults"]
+#     + df_clean.loc[missing_values, "count_children"]
+#     + df_clean.loc[missing_values, "count_babies"]
+# )
+
+# # Exporting dataframe to clean folder
+# df_clean.to_parquet(f"data-modified/users.parquet")
+```
+
+
+```python
+# Clean Inventory dataframe
+# Importing dataframe 
+df = pd.read_parquet(f"{folder_path}/inventory.parquet")
+
+#Adding np.nan to products with price equal to zero
+df["price"].mask(df["price"] == 0, np.nan, inplace=True)
+df["compare_at_price"].mask(df["compare_at_price"] == 0, np.nan, inplace=True)
+df.to_parquet(f"{folder_path}/inventory_clean.parquet")
+
+```
+
+#### Añadiendo informacion de usuarios a cada pedido
+
+
+```python
+# Merging orders dataframe
+
+df_orders = pd.read_parquet(f"{folder_path}/orders_clean.parquet")
+df_users = pd.read_parquet(f"{folder_path}/users.parquet")
+df_inventory = pd.read_parquet(f"{folder_path}/inventory_clean.parquet")
+
+df_final = df_orders.copy()
+
+
+# Left join to complete users information (count adults, count children, etc.)
+df_final = pd.merge(df_final, df_users, how="left", on=["user_id"])
+
+
+# Explode list of items whitin orders
+df_explode = df_final.explode("ordered_items")
+df_explode.rename(columns={"ordered_items": "variant_id"}, inplace=True)
+
+# Merging with inventory dataframe
+df_explode = pd.merge(df_explode, df_inventory, how="left", on=["variant_id"])
+df_explode.dropna(inplace=True)
+
+# Grouping information added
+total_price = df_explode.groupby("id")["price"].agg("sum")
+max_price_article = df_explode.groupby("id")["price"].agg("max")
+total_compare_at_price = df_explode.groupby("id")["compare_at_price"].agg("sum")
+vendors = df_explode.groupby("id")["vendor"].agg(list)
+product_types = df_explode.groupby("id")["product_type"].agg(list)
+tags = df_explode.groupby("id")["tags"].agg(list)
+
+# Adding grouped information added to final dataframe
+df_final = pd.merge(df_final, total_price, how="left", on=["id"])
+df_final = pd.merge(df_final, total_compare_at_price, how="left", on=["id"])
+df_final = pd.merge(df_final, max_price_article, how="left", on=["id"])
+
+df_final = pd.merge(df_final, vendors, how="left", on=["id"])
+df_final = pd.merge(df_final, product_types, how="left", on=["id"])
+df_final = pd.merge(df_final, tags, how="left", on=["id"])
+
+
+df_final.rename(columns={"price_x": "mean_price", "price_y": "max_article_price"}, inplace=True)
+
+# If prices are zero, add np.nan with mask
+df_final["mean_price"].mask(df_final["mean_price"] == 0, np.nan, inplace=True)
```

Comment:
Try avoiding this inplace as much as possible. They can affect other objects referencing the same dataset.